zproject/backends.py

```diff
@@ -1630,6 +1631,93 @@ def redirect_deactivated_user_to_login(realm: Realm, email: str) -> HttpResponse
     return HttpResponseRedirect(redirect_url)
 
 
+def social_auth_sync_user_attributes(
+    realm: Realm, user_profile: UserProfile | None, extra_attrs: dict[str, Any], backend: Any
+) -> int | None:
+    """
+    Syncs user attributes based on the SOCIAL_AUTH_SYNC_ATTRS_DICT setting.
+    Only supports:
+    1. Syncing the role. This is plumbed through to user creation, so can be
+       used to immediately create new users with their role set based on an attribute
+       provided by the IdP.
+    2. Syncing custom attributes. This isn't supported for user creation,
+       so they'll only be synced during the user's next login, not during
+       signup.
+    """
+    # This is only supported for SAML right now, though the design
+    # is meant to be easy to extend this to other backends if desired.
+    # Unlike with LDAP, here we can only do syncing during the authentication
+    # flow, as that's when the data is provided and we don't have a way to query
+    # for it otherwise.
+    assert backend.name == "saml"
+
+    attrs_by_backend = settings.SOCIAL_AUTH_SYNC_ATTRS_DICT.get(realm.subdomain, {})
+    profile_field_name_to_attr_name = attrs_by_backend.get(backend.name, {})
+    if not extra_attrs or not profile_field_name_to_attr_name:
+        return None
+
+    user_id = None
+    if user_profile is not None:
+        user_id = user_profile.id
+
+    custom_profile_field_name_to_value = {}
+    new_role = None
+    for field_name, attr_name in profile_field_name_to_attr_name.items():
+        if field_name == "role":
+            ROLE_NAME_TO_TYPE = {
+                "owner": UserProfile.ROLE_REALM_OWNER,
+                "administrator": UserProfile.ROLE_REALM_ADMINISTRATOR,
+                "moderator": UserProfile.ROLE_MODERATOR,
+                "member": UserProfile.ROLE_MEMBER,
+                "guest": UserProfile.ROLE_GUEST,
+            }
+            ROLE_TYPE_TO_NAME = {v: k for k, v in ROLE_NAME_TO_TYPE.items()}
```

Comment:
This mapping is duplicated with https://github.com/zulip/zulip/blob/30e20734d30a975d7cf913a1568ab6cf93d1ed76/zerver/lib/scim.py#L35-L43

Probably should be extracted to a shared place. Any proposals? Just `zerver/models/users.py`?

Reply:
`zerver/models/users.py` seems fine.